commands/products.go

```diff
@@ -0,0 +1,136 @@
+package commands
+
+import (
+	"fmt"
+	"github.com/pivotal-cf/om/api"
+	"github.com/pivotal-cf/om/models"
+	"sort"
+
+	"github.com/pivotal-cf/jhanda"
+	"github.com/pivotal-cf/om/presenters"
+)
+
+//counterfeiter:generate -o ./fakes/product_service.go --fake-name ProductService . productService
+type productService interface {
+	GetDiagnosticReport() (api.DiagnosticReport, error)
+	ListAvailableProducts() (api.AvailableProductsOutput, error)
+}
+
+type byProductName []models.ProductVersions
+
+func (a byProductName) Len() int           { return len(a) }
+func (a byProductName) Less(i, j int) bool { return a[i].Name < a[j].Name }
+func (a byProductName) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+type Products struct {
+	presenter      presenters.FormattedPresenter
+	productService productService
+	Options        struct {
+		Available bool   `long:"available" short:"a" default:"false" description:"Specify to include available products. Can be used with other options."`
+		Staged    bool   `long:"staged" short:"s" default:"false" description:"Specify to include staged products. Can be used with other options."`
+		Deployed  bool   `long:"deployed" short:"d" default:"false" description:"Specify to deployed products. Can be used with other options."`
+		Format    string `long:"format" short:"f" default:"table" description:"Format to print as (options: table,json)"`
+	}
+}
+
+func NewProducts(presenter presenters.FormattedPresenter, productService productService) Products {
+	return Products{
+		presenter:      presenter,
+		productService: productService,
+	}
+}
+
+func (sp Products) Execute(args []string) error {
+	if _, err := jhanda.Parse(&sp.Options, args); err != nil {
+		return fmt.Errorf("could not parse products flags: %s", err)
+	}
+
+	diagnosticReport, err := sp.productService.GetDiagnosticReport()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve staged and deployed products %s", err)
+	}
+
+	availableProducts, err := sp.productService.ListAvailableProducts()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve available products %s", err)
+	}
+
+	stagedProducts := diagnosticReport.StagedProducts
+	deployedProducts := diagnosticReport.DeployedProducts
+
+	sp.presenter.SetFormat(sp.Options.Format)
+
+	productVersionsCombiner := make(map[string]models.ProductVersions)
+	for _, product := range stagedProducts {
+		if _, ok := productVersionsCombiner[product.Name]; !ok {
+			productVersionsCombiner[product.Name] = models.ProductVersions{
+				Name: product.Name,
+			}
+		}
+
+		productVersions := productVersionsCombiner[product.Name]
+		productVersions.Staged = product.Version
+		productVersionsCombiner[product.Name] = productVersions
+	}
+
+	for _, product := range deployedProducts {
```

Comment:
This whole section is clear, and easy to read. If I'm trying to be nitpicky, I'd love to see these `for` loops in functions (`combineStagedProducts`, `combineAvailableProducts`, `combineDeployedProducts` or something similar)
Just to let this flow top->bottom a little faster without the for loop break

Reply:
if these functions also have the `sp` receiver like `Execute`, you can also clean up some of the earlier setup

Reply:
The only reason I didn't pull these into separate functions was the complexity of their signatures when using Intellij's extract method feature. It might not have actually needed to be that complicated, though. I just didn't dig further at that point.

Reply:
had a quick additional commit to implement this.